sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java

```diff
@@ -950,6 +1007,338 @@ public int hashCode() {
+      }
+
+      @Override
+      public int hashCode() {
+        return Objects.hash(type, value);
+      }
+    }
+
+    public static class Builder {
+      private Map<String, Option> options;
+
+      Builder(Map<String, Option> init) {
+        this.options = new HashMap<>(init);
+      }
+
+      Builder() {
+        this(new HashMap<>());
+      }
+
+      public Builder setByteOption(String optionName, Byte value) {
+        setOption(optionName, FieldType.BYTE, value);
+        return this;
+      }
+
+      public Builder setBytesOption(String optionName, byte[] value) {
+        setOption(optionName, FieldType.BYTES, value);
+        return this;
+      }
+
+      public Builder setInt16Option(String optionName, Short value) {
+        setOption(optionName, FieldType.INT16, value);
+        return this;
+      }
+
+      public Builder setInt32Option(String optionName, Integer value) {
+        setOption(optionName, FieldType.INT32, value);
+        return this;
+      }
+
+      public Builder setInt64Option(String optionName, Long value) {
+        setOption(optionName, FieldType.INT64, value);
+        return this;
+      }
+
+      public Builder setDecimalOption(String optionName, BigDecimal value) {
+        setOption(optionName, FieldType.DECIMAL, value);
+        return this;
+      }
+
+      public Builder setFloatOption(String optionName, Float value) {
+        setOption(optionName, FieldType.FLOAT, value);
+        return this;
+      }
+
+      public Builder setDoubleOption(String optionName, Double value) {
+        setOption(optionName, FieldType.DOUBLE, value);
+        return this;
+      }
+
+      public Builder setStringOption(String optionName, String value) {
+        setOption(optionName, FieldType.STRING, value);
+        return this;
+      }
+
+      public Builder setDateTimeOption(String optionName, ReadableDateTime value) {
+        setOption(optionName, FieldType.DATETIME, value);
+        return this;
+      }
+
+      public Builder setBooleanOption(String optionName, Boolean value) {
+        setOption(optionName, FieldType.BOOLEAN, value);
+        return this;
+      }
+
+      public <V> Builder setArrayOption(String optionName, FieldType arrayType, List<V> value) {
+        setOption(optionName, arrayType, value);
+        return this;
+      }
+
+      public <K, V> Builder setMapOption(String optionName, FieldType mapType, Map<K, V> value) {
+        setOption(optionName, mapType, value);
+        return this;
+      }
+
+      public Builder setRowOption(String optionName, Row value) {
+        setOption(optionName, FieldType.row(value.getSchema()), value);
+        return this;
+      }
+
+      public Builder setOption(String optionName, FieldType fieldType, Object value) {
+        if (value == null) {
+          removeOption(optionName);
+        } else {
+          options.put(
+              optionName, new Option(fieldType, ValueUtils.verify(value, fieldType, optionName)));
```

Comment:
Would it be possible to store the options values as a `RowWithStorage` so we can re-use Row's verification rather than making it public in `ValueUtils`?

Reply:
Not really. As options can be primitive types as well, Row is just one of the possibilities of a type of an option.

Reply:
I meant that all of the options could be stored in a Row rather than in a HashMap

Reply:
Mmm... I'm not sure. I think it would set too much limits on the options name. It's important that option names from proto and avro can be mapped into the options. They generally will contain dot's. Don't think that's allowed.

Reply:
Yeah I think we disallow dots in field names (but I don't think these limitations are documented anywhere..), since we want to be able to use "." to access nested fields in a FieldAccessDescriptor unambiguously.

What about mapping the dots in proto options to a level of nesting? so if you have an option named "foo.bar" with type STRING and "baz" with type INT64 on the same field, we would map them to Options with schema <foo=STRUCT<bar=STRING>, baz=INT64>

Reply:
this feels like a hack, giving a special meaning and restructuring options just because they have a dot in them. What about URN's? This is also something that should be allowed. What are we going todo with this, also restructure? Or allow it. Replacing the HashMap by a Row is bad in my option for the following reason:
-  makes the name handling complex, now they are just a key
- Rows are now used in as value container for complex structures, they are accessed by simple arbitrary string keys. This feels natural, but hacking the Option HashMap into a Row feels bad. Making to code far more complex that it needs to be.

Reply:
Is it really that unnatural? My understanding (based on my quick look at the [docs](https://developers.google.com/protocol-buffers/docs/proto#customoptions)) is that protobuf options are themselves described by protobuf message descriptors. So they have a structure that we are already mapping to a nested Beam schema in another context. If we just used the protobuf schema provider to generate a schema from the option's descriptor(s), then any dot in the name would be referencing a level of nesting, not a literal character in the field name, wouldn't it?

I see your point about URNs, they usually define some namespaces separated by `:` characters, is this a slippery slope and we need to nest them as well? But that wouldn't be necessary since we don't have any issue with `:` in field names (that I'm aware of).

As far as code complexity - my hope was that this would actually reduce code complexity, since we can re-use a lot of existing infrastructure: RowWithStorage, Schema, FieldAccessDescriptor, and the protobuf SchemaProvider.

To be clear, I'm ok with keeping the options as-is. I just thought this may be a way to re-use more code. But you know the problem you're trying to solve better than I do so I'll defer to you if you think this doesn't make sense.

Reply:
If you look closer to the proto doc you will see that the options are defined as extension points. You could argue that they are indeed extending messages and they are analog to a Schema of a Row, but then you have to drag the complete complexity of making extendable Schema's into the mix (alias extension in Proto). The implementation with the hashmap is closer to the usage of the Proto Java API, where you do getField().getOption("").getField(<extension>)... 

This PR is reusing a lot of the Row implementation though, as options are actually Fields (that's being reused) certainly RowWithStorage and Schema are used as  store for Rows in an option. If you look at the Proto implementation of options you will see how they are implemented. (see: https://github.com/apache/beam/pull/10529)